### ISSUE TITLE
Pyside2 & Pyqt5 xi frontend

### ISSUE THREAD

**Issue by brupelo** [xi-editor/xi-editor#1132]
Hello, I'll try to be short here as I'm texting from the phone, in terms of features and speed... how would you compare xi vs sublime? Assuming the answer is positive...

How long would do you think it'd take to create a xi frontend for pyqt or pyside? I know there is already something similar who's been written for c++ and qt but it hasn't been mantained for a while.

If the task is not too long, let's say max 2 or 3 days working full time i'd be willing to take it.

But first of all, I'd like to know what to expect from xi in comparison to sublime text. I've used scintilla on my projects for very years and it's is really limited in comparison to sublime so i'd like to replace it with something better

That said, if I was going to take this project I'd need some expert from core xi team providing me the right guidance to evaluate how long this project could take.

I've got little experience with rust, mostly of it wrapping rust into python but i'm definitely quite an expert on python,c++,qt,pyqt,pyside2 😁

If anybody with expertise on xi want this to happen, please let me know, thanks in advance.

**Comment by brupelo** [comment]
Closing this issue as i've tested the [windows frontend](https://github.com/xi-editor/xi-win) and it's not exactly what i was expecting after reading all the nice introduction ;)

**Comment by jansol** [comment]
The windows frontend is also the one that has received the least attention (read: practically none)... If you want to get a proper idea of the state of xi-editor you basically have to try out xi-mac. Or gxi, although it might lag slightly behind.

**Comment by brupelo** [comment]
@jansol So... basically you're saying the windows frontend isn't exposing all the "power" from xi core? To me one of the most importants features from Sublime is how it handles multiple selections. Scintilla also handles multiple selection but not in the best way, as Sublime does. The way selection works in the windows frontend is buggy and bad and it contained a lot of other bugs.

Anyway, I don't have a mac to try out xi-mac, that's unfortunate and not sure gxi would work on windows... I'll try to find some videos on youtube to get a proper idea of the state of xi-editor though

**Comment by LiHRaM** [comment]
@brupelo That is exactly right. xi-win is in _very_ early stages of development, and is not really being actively developed now, as its underlying frameworks are being worked on instead. [druid](https://github.com/xi-editor/druid/) and [piet](https://github.com/linebender/piet). Although the progress in those frameworks is exciting, it does mean that xi-win is not the top priority.

**Comment by brupelo** [comment]
@LiHRaM I've decided to spend some hours this morning to toy around with xi-editor & pyqt, one thing I really like was that building xi-core was painless and the fact there is a nice separation between frontend & backend, so it shouldn't be too hard having something up&running soon.

Right now I'm just trying to figure out how initialize xi-core server properly though :) ===> [https://github.com/brupelo/pyxiqt/blob/master/pyxiqt/main.py](https://github.com/brupelo/pyxiqt/blob/master/pyxiqt/main.py)

As you can see below the output won't always be the same:

![showcase](https://dl.dropboxusercontent.com/s/p2udkjmu148xta8/ConEmu_2019-03-12_11-30-02.png)

Anyway, let's see :)

**Comment by brupelo** [comment]
Here's what I've got right [now](https://github.com/brupelo/pyxiqt/blob/65265dae2e59b7ca3474974937b79b5146df7039/pyxiqt/main.py). I've got stuck with this error:

    [2019-03-12][15:55:07][xi_core][INFO] Logging with fern is set up
    [2019-03-12][15:55:07][xi_core][INFO] Writing logs to: C:\Users\KneDa\AppData\Local\xi-core\xi-core.log
    Xi Server listening...
    b'{"method": "new_view", "file_path": "D:\\\\sources\\\\personal\\\\python\\\\apps\\\\pyqt\\\\pyxiqt\\\\pyxiqt\\\\main.py"}'
    b'31215550]xi_core] UnknownRequest(""00)'
    [2019-0-][::7[][ERRORxi-core exited with error:
    Error(unknown variant `new_view`, expected one of `edit`, `plugin`, `close_view`, `save`, `set_theme`, `client_started`, `modify_user_config`, `tracing_config`, `save_trace`, `set_language`, line: , column: )

Could anyone tell me what's wrong either on this fragment [https://github.com/brupelo/pyxiqt/blob/master/pyxiqt/main.py#L105-L111](https://github.com/brupelo/pyxiqt/blob/master/pyxiqt/main.py#L105-L111) or this one [https://github.com/brupelo/pyxiqt/blob/master/pyxiqt/main.py#L123-L126](https://github.com/brupelo/pyxiqt/blob/master/pyxiqt/main.py#L123-L126)?

**Comment by LiHRaM** [comment]
The first thing the core expects is a notification that the client has started. Let me find the relevant docs for you.

**Comment by brupelo** [comment]
Thanks for your fast response, when I'm back home tonight I'll check it out. Next thing I'll do will be probably wrap the whole Xi Api followings the docs... Right now im using the json rpc server but if I like how the editor looks like and the type features it offers I guess I'll end up using xi core features wrapped as a python module and compare whether that'd be a much faster approach.

Btw, any idea, suggestion, consideration.. just let me know. Everything helps as I'd like to check sooner than later whether xi editor will be a good drop in replacement for my scintilla widgets. Ty

**Comment by jansol** [comment]
@brupelo you might want to hop on [the xi-editor zulip group](https://xi.zulipchat.com/) (auth with github) for more in-depth discussion.

**Comment by brupelo** [comment]
@jansol Thanks to let me know but I definitely prefer using the good old IRC, it seems there is a #xi channel at moznet and today I wanted to give it a shot to that channel but when tried to register my nickname (#xi only accept registered users) any confirmation mail appeared at my old mixmail spam account :) and now I need to wait 10 hours to register again with a different mail account, tomorrow I'll be there (hopefully :))

@LiHRaM Thanks for your commit suggestion, that was helpful and it just worked (you can check my last commit) and as you said other type of information like languages & themes were returned by xi.

That said, before thinking how I'm gonna design my data structures to present the information retrieved by xi in pyqt I'd like to have very clear how the whole bunch of messages are used in both FRONTEND->BACKEND and BACKEND->FRONTEND communications . 

I've been reading the official docs, ie: [https://xi-editor.io/docs/frontend-protocol.html#from-front-end-to-back-end](https://xi-editor.io/docs/frontend-protocol.html#from-front-end-to-back-end) and I must to say those docs are confusing in some places. As a frontend developer I just really care about the whole (and complete) public API. A list of all specific messages with the proper format would be more than good enough, Actually, I don't need any unnecessary irrelevant information other than that. Something similar to this [https://www.sublimetext.com/docs/3/api_reference.html](https://www.sublimetext.com/docs/3/api_reference.html) is what I call a good presentation of information.

That said, could anyone clarify a little bit about it some of the below topics? I've got no clue what's the format I need to use to query the Xi server:

[other-movement-and-deletion-commands](https://xi-editor.io/docs/frontend-protocol.html#other-movement-and-deletion-commands)
[transformations](https://xi-editor.io/docs/frontend-protocol.html#transformations)
[number-transformations](https://xi-editor.io/docs/frontend-protocol.html#number-transformations)
[recording](https://xi-editor.io/docs/frontend-protocol.html#recording)
[language-support-oriented-features-in-edit-namespace](https://xi-editor.io/docs/frontend-protocol.html#language-support-oriented-features-in-edit-namespace)

Also, some parts from [https://xi-editor.io/docs/frontend-protocol.html#from-back-end-to-front-end](https://xi-editor.io/docs/frontend-protocol.html#from-back-end-to-front-end) are confusing but i guess in this case I'll need to read carefully the blob of text provided...

In fact, sometimes you use "method" in the docs, in other cases you don't, it's not really homogenous and uniform documentation overall... sometimes even looks a bit random :D

**Comment by Cogitri** [comment]
Yeah, we definitely need to work on the docs, I shall so that once I'm back home from holidays
